### PR TITLE
Output dim tag of torch.full

### LINKED ISSUE #114: Output dim tag of torch.full
In #113, it shows that the dim tag of the output shape of `torch.full` is not set correctly. An example to reproduce the issue looks like this:
```
def test_full_out_dim_tag():
  n_batch, n_feat = 3, 5

  def model_func(wrapped_import, inputs: torch.Tensor):
    if typing.TYPE_CHECKING or not wrapped_import:
      import torch
    else:
      torch = wrapped_import("torch")
    full = torch.full((inputs.shape[0],), 42)
    arange = torch.arange(full .shape[0])
    return arange + full 

  rnd = numpy.random.RandomState(42)
  x = rnd.normal(0., 1., (n_batch, n_feat)).astype("float32")
  verify_torch_and_convert_to_returnn(model_func, inputs=x)
```
which fails in `pytorch_to_returnn/torch/nn/modules/operator.py`, line 594, in `_unify_tensor_axes_returnn_meta` with
```
    line: assert all(dim.dimension == d.dimension for d in dims_for_axis if d is not None and d.dimension != 1), (
            f"invalid input {x} axis {i} dim {dim}")
    locals:
      all = <builtin> <built-in function all>
      dim = <local> Dim{'Range:range'(3)}
      dim.dimension = <local> 3
      d = <not found>
      d.dimension = <not found>
      dims_for_axis = <local> [Dim{'Range:range'(3)}, Dim{B}]
AssertionError: invalid input <Tensor name:? tensor:('static_dim'(3)(3),) returnn_data:'Cast_output' [F|'Range:range'(3)] axes id> axis 0 dim Dim{'Range:range'(3)}
```
(I'll add a full stack trace and log later).

We can see that the dim tag in the output shape of `torch.full` is wrong, but in `returnn_data` it's still correct. When creating a range based on that, the wrong dim tag of the output shape is also used on RETURNN side, which will lead to the error when trying to add both tensors.
```
      arange = <local> <Tensor name:? tensor:('static_dim'(3)(3),) returnn_data:'Range_output' [F|'Range:range'(3)] axes id>
      full = <local> <Tensor name:? tensor:('static_dim'(3)(3),) returnn_data:'FullStatic_const' [B] axes id>
```

### PR DESCRIPTION
**Pull request by vieting**
Fix #114

### CODE REVIEW

**Comment by albertz** [comment]
I don't remember the details anymore. `limit.is_defined` makes only sense for static dims, right? And if we have `limit.returnn_naming_entry.is_size_value`, we don't need that (we don't need `limit = limit.numpy()`).

**Comment by albertz** [comment]
I'm not sure if we have `limit.returnn_naming_entry.is_size_value` here though. Effectively, `limit` is also not the size itself but the max of the seq lengths. Which should be the same though. We probably need some extra logic to set this correctly then. E.g. when you do `torch.max(lengths)`.

**Comment by vieting** [comment]
> I don't remember the details anymore. limit.is_defined makes only sense for static dims, right? And if we have limit.returnn_naming_entry.is_size_value, we don't need that (we don't need limit = limit.numpy()).

Yes, I agree.

> I'm not sure if we have limit.returnn_naming_entry.is_size_value here though. Effectively, limit is also not the size itself but the max of the seq lengths. Which should be the same though. We probably need some extra logic to set this correctly then. E.g. when you do torch.max(lengths).

Right, it's `None` here. I'll look into the `torch.max`.

**Comment by vieting** [comment]
> I'm not sure if we have `limit.returnn_naming_entry.is_size_value` here though. Effectively, `limit` is also not the size itself but the max of the seq lengths. Which should be the same though. We probably need some extra logic to set this correctly then. E.g. when you do `torch.max(lengths)`.

So for `Reduce` with `mode="max"`, we would need to check if we can already know what that is?

**Comment by albertz** [comment]
> > I'm not sure if we have `limit.returnn_naming_entry.is_size_value` here though. Effectively, `limit` is also not the size itself but the max of the seq lengths. Which should be the same though. We probably need some extra logic to set this correctly then. E.g. when you do `torch.max(lengths)`.
> 
> So for `Reduce` with `mode="max"`, we would need to check if we can already know what that is?

What fails now with your change is e.g. `test_arange_from_lengths`. In there, we have this code:
```
batch_dim, time_dim = inputs.shape
lengths = torch.full((batch_dim,), time_dim)
arange = torch.arange(torch.max(lengths))
```
In practice, it is a question where `lengths` comes from. I think so far we never really attached the sequence lengths to the original tensor in any way, or did we?

In this synthetic unrealistic example, we could infer that `torch.max(lengths)` refers to `time_dim` which has the correct `is_size_value`. Is the same actually, we maybe can even return the same tensor object.

We should fix this example such that the tests pass. Which probably should not be too difficult. Or otherwise the tests somehow needs to be rewritten.

The more relevant question however is: How will it look in practice, i.e. where do the `lengths` come from, and how are they connected to `inputs`?

Maybe you could just also set `is_size_value` of `lengths` here in this example (and in general, however you would get `lengths`). This is not just a shape dim (scalar) but the actual lengths but we could treat it in the same way. That would then be easy to detect in `torch.max`. Not sure if this is a good solution or whether this causes other problems elsewhere.

**Comment by vieting** [comment]
I have pushed a fix, but it's very specific. In `Reduce` with `mode in ["max", "min", "mean"]`, it checks if the input comes from `FullStatic` and sets `is_size_value` for the tensor entry if the `fill_value` is a `SizeValue`.

**Comment by vieting** [comment]
> The more relevant question however is: How will it look in practice, i.e. where do the lengths come from, and how are they connected to inputs?

I don't remember the details, but I think the motivation for the test was `test_packed_sequence_2` or, more generally, `torch.nn.utils.rnn.pad_packed_sequence`.

**Comment by albertz** [comment]
> I have pushed a fix, but it's very specific. In `Reduce` with `mode in ["max", "min", "mean"]`, it checks if the input comes from `FullStatic` and sets `is_size_value` for the tensor entry if the `fill_value` is a `SizeValue`.

But it only makes sense for `max`, or not?

**Comment by albertz** [comment]
> > The more relevant question however is: How will it look in practice, i.e. where do the lengths come from, and how are they connected to inputs?
> 
> I don't remember the details, but I think the motivation for the test was `test_packed_sequence_2` or, more generally, `torch.nn.utils.rnn.pack_sequence`.

But also there, the same question: Where do the `lengths` come from?

**Comment by vieting** [comment]
> > I have pushed a fix, but it's very specific. In `Reduce` with `mode in ["max", "min", "mean"]`, it checks if the input comes from `FullStatic` and sets `is_size_value` for the tensor entry if the `fill_value` is a `SizeValue`.
> 
> But it only makes sense for `max`, or not?

If it is `FullStatic`, then all values in the tensor are the same, so max = min = mean, ot not?

**Comment by vieting** [comment]
> > > The more relevant question however is: How will it look in practice, i.e. where do the lengths come from, and how are they connected to inputs?
> > 
> > 
> > I don't remember the details, but I think the motivation for the test was `test_packed_sequence_2` or, more generally, `torch.nn.utils.rnn.pack_sequence`.
> 
> But also there, the same question: Where do the `lengths` come from?

In `pack_sequence` we have
```
    batch_dim = sequences.shape[torch_axis_from_returnn_axis[tensor_entry.returnn_data.batch_dim_axis]]
    time_dim = sequences.shape[torch_axis_from_returnn_axis[tensor_entry.returnn_data.time_dim_axis]]
    lengths = full((batch_dim,), time_dim)
    # batch_first is always True because we assume batch-time-major tensor, see above
    return pack_padded_sequence(sequences, lengths, enforce_sorted=enforce_sorted, batch_first=True)
```
so `lengths` is created by `full`. Then in `pack_padded_sequence`, the `lengths` are converted to `batch_sizes` via `_batch_sizes_from_lengths` which uses `torch.arange(torch.max(lengths))`. So this is pretty much the same as what we test in `test_arange_from_lengths`.

**Comment by albertz** [comment]
> > > I have pushed a fix, but it's very specific. In `Reduce` with `mode in ["max", "min", "mean"]`, it checks if the input comes from `FullStatic` and sets `is_size_value` for the tensor entry if the `fill_value` is a `SizeValue`.
> > 
> > But it only makes sense for `max`, or not?
> 
> If it is `FullStatic`, then all values in the tensor are the same, so max = min = mean, ot not?

I don't understand why this is relevant? Normally you would never use `FullStatic` in practice for the seq lengths. The seq lengths are usually never all the same. So in the general case, only `max` is correct.

**Comment by albertz** [comment]
> In `pack_sequence` we have
> 
> ```
>     batch_dim = sequences.shape[torch_axis_from_returnn_axis[tensor_entry.returnn_data.batch_dim_axis]]
>     time_dim = sequences.shape[torch_axis_from_returnn_axis[tensor_entry.returnn_data.time_dim_axis]]
>     lengths = full((batch_dim,), time_dim)
>     # batch_first is always True because we assume batch-time-major tensor, see above
>     return pack_padded_sequence(sequences, lengths, enforce_sorted=enforce_sorted, batch_first=True)
> ```
> 
> so `lengths` is created by `full`. Then in `pack_padded_sequence`, the `lengths` are converted to `batch_sizes` via `_batch_sizes_from_lengths` which uses `torch.arange(torch.max(lengths))`. So this is pretty much the same as what we test in `test_arange_from_lengths`.

But why is this relevant? This code is obviously wrong, and you would use it in practice. You do not have the same seq lengths in practice.

File: pytorch_to_returnn/torch/nn/modules/operator.py [hunk 1]
```diff
@@ -516,6 +516,23 @@ def create_returnn_layer_dict(self, input: Tensor) -> Dict[str, Any]:
...
+    if self.mode in ["max", "min", "mean"]:
+      # handle special case where input to Reduce is FullStatic. Here, we might already know the output and in case it
+      # is a SizeValue, we can set it for the TensorEntry
+      naming = Naming.get_instance()
+      assert len(inputs_flat) == 1
+      input_entry = naming.tensors[inputs_flat[0]]
+      from .variable import FullStatic
+      if len(input_entry.output_from_calls) == 1:
+        mod = input_entry.output_from_calls[0].module.module
+        if isinstance(mod, FullStatic):
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/operator.py:530]: You should not have any check for `FullStatic` here. `FullStatic` should be totally irrelevant in this function.

File: pytorch_to_returnn/torch/nn/modules/operator.py [hunk 2]
```diff
@@ -516,6 +516,23 @@ def create_returnn_layer_dict(self, input: Tensor) -> Dict[str, Any]:
     axes = [self._get_input_axis_to_returnn(input, axis) for axis in axes]
     return {"class": "reduce", "mode": self.mode, "axes": axes, "from": self._get_input_layer_name(input)}
 
+  def make_output_tensor_from_returnn(self, inputs_flat: List[Tensor], layer: LayerBase) -> Tensor:
+    tensor = super(Reduce, self).make_output_tensor_from_returnn(inputs_flat, layer)
+    if self.mode in ["max", "min", "mean"]:
+      # handle special case where input to Reduce is FullStatic. Here, we might already know the output and in case it
```
Comment (albertz) [inline pytorch_to_returnn/torch/nn/modules/operator.py:522]: You should not have a special case for this.

**Comment by albertz** [comment]
> Maybe you could just also set `is_size_value` of `lengths` here in this example (and in general, however you would get `lengths`). This is not just a shape dim (scalar) but the actual lengths but we could treat it in the same way. That would then be easy to detect in `torch.max`. Not sure if this is a good solution or whether this causes other problems elsewhere.

Did you try this?

I.e. this would be a small extra logic in `FullStatic`.

And in `Reduce`, you would just check for `is_size_value` in case of `max` (and only `max`).

**Comment by vieting** [comment]
> > Maybe you could just also set `is_size_value` of `lengths` here in this example (and in general, however you would get `lengths`). This is not just a shape dim (scalar) but the actual lengths but we could treat it in the same way. That would then be easy to detect in `torch.max`. Not sure if this is a good solution or whether this causes other problems elsewhere.
> 
> Did you try this?
> 
> I.e. this would be a small extra logic in `FullStatic`.
> 
> And in `Reduce`, you would just check for `is_size_value` in case of `max` (and only `max`).

I pushed something, is that what you mean?

**Comment by albertz** [comment]
> I pushed something, is that what you mean?

Yes, looks good.

**Review by albertz — APPROVED** [review]

**Pull request merged**